framework/include/base/MooseObjectStorage.h

```diff
@@ -0,0 +1,398 @@
+    for (std::set<SubdomainID>::const_iterator it = ids.begin(); it != ids.end(); ++it)
+    {
+      _all_block_objects[tid][*it].push_back(object);
+      if (enabled)
+        _active_block_objects[tid][*it].push_back(object);
+    }
+  }
+}
+
+
+template<typename MooseObjectType>
+bool
+MooseObjectStorage<MooseObjectType>::hasActiveBlockObjects(THREAD_ID tid/* = 0*/) const
+{
+  checkThreadID(tid);
+  bool has_active_block_objects = false;
+  typename std::map<SubdomainID, std::vector<MooseSharedPointer<MooseObjectType> > >::const_iterator it;
+  for (it = _active_block_objects[tid].begin(); it != _active_block_objects[tid].end(); ++it)
+    has_active_block_objects |= !(it->second.empty());
+  return has_active_block_objects;
+}
+
+
+template<typename MooseObjectType>
+bool
+MooseObjectStorage<MooseObjectType>::hasActiveBlockObjects(SubdomainID id, THREAD_ID tid/* = 0*/) const
+{
+  checkThreadID(tid);
+  typename std::map<SubdomainID, std::vector<MooseSharedPointer<MooseObjectType> > >::const_iterator iter = _active_block_objects[tid].find(id);
+  if (iter != _active_block_objects[tid].end() && !iter->second.empty())
+    return true;
+  return false;
+}
+
+
+template<typename MooseObjectType>
+bool
+MooseObjectStorage<MooseObjectType>::hasActiveBoundaryObjects(THREAD_ID tid/* = 0*/) const
+{
+  checkThreadID(tid);
+  bool has_active_boundary_objects = false;
+  typename std::map<BoundaryID, std::vector<MooseSharedPointer<MooseObjectType> > >::const_iterator it;
+  for (it = _active_boundary_objects[tid].begin(); it != _active_boundary_objects[tid].end(); ++it)
+    has_active_boundary_objects |= !(it->second.empty());
+  return has_active_boundary_objects;
+}
+
+
+template<typename MooseObjectType>
+bool
+MooseObjectStorage<MooseObjectType>::hasActiveBoundaryObjects(BoundaryID id, THREAD_ID tid/* = 0*/) const
+{
+  checkThreadID(tid);
+  typename std::map<BoundaryID, std::vector<MooseSharedPointer<MooseObjectType> > >::const_iterator iter = _active_boundary_objects[tid].find(id);
+  if (iter != _active_boundary_objects[tid].end() && !iter->second.empty())
+    return true;
+  return false;
+}
+
+
+template<typename MooseObjectType>
+void
+MooseObjectStorage<MooseObjectType>::updateActive(THREAD_ID tid /*= 0*/)
+{
+  checkThreadID(tid);
+
+  updateActiveHelper(_active_objects[tid], _all_objects[tid]);
+
+  {
+    typename std::map<SubdomainID, std::vector<MooseSharedPointer<MooseObjectType> > >::const_iterator it;
+    for (it = _all_block_objects[tid].begin(); it != _all_block_objects[tid].end(); ++it)
+      updateActiveHelper(_active_block_objects[tid][it->first], it->second);
+  }
+
+
+  {
+    typename std::map<BoundaryID, std::vector<MooseSharedPointer<MooseObjectType> > >::const_iterator it;
+    for (it = _all_boundary_objects[tid].begin(); it != _all_boundary_objects[tid].end(); ++it)
+      updateActiveHelper(_active_boundary_objects[tid][it->first], it->second);
```

Comment:
Only a few classes can be both ("dual_restrictable"), most can be one or the other. I am for re-factoring this, but this is another issue: https://github.com/idaholab/moose/issues/4214